### ISSUE TITLE
asyncio.create_task does not accept anext of asynchronous generator

### ISSUE THREAD

**Issue by hurryabit** [microsoft/pyright#9029]
The type Pyright infers for asynchronous generators appears to be too general for them to work with `asyncio.create_task`. The following example illustrates the issue:

```python
import asyncio

async def gen():
    yield 42

async def main():
    await asyncio.create_task(anext(gen()))

asyncio.run(main())
```

This code runs just fine and prints 42. However, the call to `create_task` gets flagged with the following type error:

```
Argument of type "Awaitable[Literal[42]]" cannot be assigned to parameter "coro" of type "_CoroutineLike[_T@create_task]" in function "create_task"
  "Awaitable[Literal[42]]" is not assignable to "Coroutine[Any, Any, _T@create_task]"
```

The type Pyright infers for `gen()` is `AsyncGenerator[Literal[42], Any]`. typeshed's [definition of `AsyncGenerator`](https://github.com/python/typeshed/blob/eb7df6d11873914895f87fa4d183be00e94a3949/stdlib/typing.pyi#L542) starts as follows:

```python
class AsyncGenerator(AsyncIterator[_YieldT_co], Generic[_YieldT_co, _SendT_contra]):
    def __anext__(self) -> Awaitable[_YieldT_co]: ...
```

Since `create_task` expects an actual coroutine and not just any awaitable, the example above does not type check. In reality, `anext(gen())` is a coroutine though and hence `create_task` is happy at runtime.

**Comment by erictraut** [comment]
I think pyright is working correctly here.

I'm not sure what change you're proposing. Could you elaborate? Are you suggesting that the typeshed definitions are incorrect?

**Comment by hurryabit** [comment]
Looking into this further, it seems that the documentation of [`collections.abc.AsyncGenerator`](https://docs.python.org/3/library/collections.abc.html#collections.abc.AsyncGenerator) is ambiguous:

> ABC for [asynchronous generator](https://docs.python.org/3/glossary.html#term-asynchronous-generator) classes that implement the protocol defined in [PEP 525](https://peps.python.org/pep-0525/) and [PEP 492](https://peps.python.org/pep-0492/).

Depending on how you put the emphasis, this allows for two interpretations:

1. The [protocol defined in the PEPs](https://peps.python.org/pep-0525/#support-for-asynchronous-iteration-protocol) requires `__anext__` to return an awaitable, which is not necessarily a coroutine. If the purpose of `AsyncGenerator` is to capture this, then typeshed is correc. In this case, Pyright should infer a more specific type for asynchronous generators defined using the `async def`/`yield` syntax which reflects that `__anext__` returns a coroutine for them.

2. The glossary entry for [asynchronous generator](https://docs.python.org/3/glossary.html#term-asynchronous-generator) says

    > [...] it looks like a coroutine function defined with `async def` except that it contains `yield` expressions [...]

    If the purpose of `AsyncGenerator` is to type asynchronous generators defined using this syntax, then Pyright is correct. In this case, the definition on typeshed is too general and should read something like

    ```python
    class AsyncGenerator(AsyncIterator[_YieldT_co], Generic[_YieldT_co, _SendT_contra]):
        def __anext__(self) -> Coroutine[Any, Any, _YieldT_co]: ...

    ```

    instead.

Given how the `asend` method of `AsyncGenerator` is typed, it seems that (2) is the correct interpretation. Can you confirm this interpretation or do you know who could? (Apologies, for asking such basic questions, I'm pretty new to the Python community.)

**Comment by erictraut** [comment]
The maintainers of the [typeshed project](https://github.com/python/typeshed) are more knowledgeable about this behavior. I recommend filing an issue there if you think that the typeshed stubs are incorrect.

I confirmed that pyright's behavior is consistent with mypy, another Python static type checker. Mypy doesn't infer return types, but it does check consistency of return type annotations. It insists that the return type of `gen` must be an `AsyncGenerator` or one of its supertypes, which includes `AsyncIterator`, `AsyncIterable` and `object.

I think pyright is behaving correctly here, so I'm going to close the issue. If you discover more details that seem to contradict this conclusion, please post back here, and I'd be happy to revisit the decision.

**Comment by hurryabit** [comment]
I don't think this is as clear cut as you suggest but this project is probably also not the right place to discuss the issue. I've opened a topic for this issue on dissuss.python.org: https://discuss.python.org/t/types-for-asynchronous-generators-too-general/64515
